### ISSUE TITLE
Training_loss = NaN of deep speaker!

### ISSUE THREAD

**Issue by shubham-iitg-ece** [philipperemy/deep-speaker#7]
While training the model, during 231st batch, training loss = 0.10000002384185791, but next batch onwards training loss = NaN. What could be possible reason for this? What is the fix for this problem. It doesn't seem to me as problem of divergence, since training loss has abruptly changed from 0.1 to NaN. Some clarifications please.

**Comment by jageshmaharjan** [comment]
@shubhamaryan , I had the same problem. may be we should try to change the hyper parameters. such as learning rate. or try using different optimization algorithms, such as  "sgd, momentum, rms", however the best one is "adam (adaptive momentum)" [used here].

If u figure it out let me know as well.

**Comment by shubhamaryan** [comment]
If it would be a problem of learning rate or optimizer, then training loss would have diverged gradually from 0.1. Secondly, there is no problem with input as well, since I have matched the log file and audio dataset and all those files exist. Hence, I am not able to conclude any further what the problem may be. If we could get some help from the developer?

**Comment by philipperemy** [comment]
@shubhamaryan for now the implementation is not clearly operational. It's still a bit buggy. Especially it's very hard to train a triplet loss. I'm thinking of doing some pre-training like what they did in the paper.

**Comment by philipperemy** [comment]
Your contributions are very welcome guys @shubhamaryan @jageshmaharjan !

**Comment by shubhamaryan** [comment]
@philipperemy @jageshmaharjan 

I changed the learning rate of Adam optimizer to 0.01 from the default value of 0.001 & regularization strength λ = 0.001, in the code, it is used as 0.0001. I also increased the sampling rate to around 1200. 

Initially, we were encountering a NaN training loss after around 230 batches. By doing above modification, I  extended training till 400 batches. I stopped the training after it, as the loss became constant after 360 batches.

**Comment by philipperemy** [comment]
@shubhamaryan Thanks.
But do you think we can have something meaningful?
I fear that the triplet loss is very hard to optimize and requires pre-training on a supervised learning.
In the paper they did something similar.

**Comment by mengxingzhi** [comment]
stub_targets = np.random.uniform(size=(x.shape[0], 1)), why is the label is selected randomly? Shouldn't it to be the true label of the speech?

**Comment by linhdvu14** [comment]
Speech label is not used in triplet loss computation. All that's needed is triplets of (anchor, positive, negative) embeddings, all of which are contained in x. stub_targets is only there to conform to API requirement.

**Comment by mengxingzhi** [comment]
@linhdvu14 thank you for your answer. I come up with another two problems: the config of my server is 12g, but  the batchsize can't be bigger than 3. Also my loss is over 70, can't be convergent. waiting for guidance, thank you!

**Comment by philipperemy** [comment]
Try again with the new master. I pushed the latest code. It's working correctly now!
